File: LeetCode - Top Interview 150/173.py
# 173. Binary Search Tree Iterator
# https://leetcode.com/problems/binary-search-tree-iterator/


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.inorder_values = []

        def get_inorder_values(node: Optional[TreeNode]):
            if not node:
                return

            get_inorder_values(node.left)
            self.inorder_values.append(node.val)
            get_inorder_values(node.right)

        get_inorder_values(root)

        self.current_index = -1

    def next(self) -> int:
        self.current_index += 1

        return self.inorder_values[self.current_index]

    def hasNext(self) -> bool:
        return self.current_index + 1 < len(self.inorder_values)

File: LeetCode - Algorithm/Binary Tree/173.js
// 173. Binary Search Tree Iterator
// https://leetcode.com/problems/binary-search-tree-iterator/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 */
var BSTIterator = function (root) {
  this.nodes = [];

  getInorder(root, this.nodes);
};

const getInorder = (node, nodes) => {
  if (!node) return;

  getInorder(node.left, nodes);
  nodes.push(node.val);
  getInorder(node.right, nodes);
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function () {
  return this.nodes.shift();
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
  return this.nodes.length > 0;
};
